Implement support for ManagedCertificate CRD

krzykwas opened the pull request:
This change adds to Ingress support for ManagedCertificate CRD to integrate with https://github.com/GoogleCloudPlatform/gke-managed-certs

krzykwas commented:
I signed it

krzykwas commented:
assign @bowei

krzykwas commented:
/assign @bowei

rramkumar1 commented:
/ok-to-test

krzykwas commented:
/assign @rramkumar1

rramkumar1 commented on cmd/glbc/main.go:
```diff
@@ -82,6 +83,11 @@ func main() {
 		glog.Fatalf("Failed to create kubernetes client: %v", err)
 	}
 
+	mcrtClient, err := managedcertificatesclient.NewForConfig(kubeConfig)
```
Can you add a short comment here stating that the ManagedCerts CRD is created + updated by another component.

krzykwas replied:
Done. Actually it is created by a user, not by managed-certificate-controller, but the status is set by that controller.

rramkumar1 replied:
IIUC, the CRD is created by the controller and the CR (custom resource) is created by the user?

krzykwas replied:
I changed the comment slightly, it's not CRD status but CR status.

My idea was to deploy the CRD via kubectl create -f (or something similar) together with managed-certificate-controller in GKE configure/startup scripts whenever managed-certificate-controller gets included in ingress controller manifest (so in alpha clusters for now). Is it a proper approach?

Users create CRs to say „I want to have ssl for example.com”. managed-certificate-controller creates SslCertificate based on CR and sets its status accordingly. Ingress reads the status and adds SslCertificate to TargetProxy.

rramkumar1 commented on pkg/controller/controller_test.go:
```diff
@@ -43,6 +44,13 @@ var (
 
+type recorderProducerMock struct {
+}
+
```
Can you move this mock to pkg/events.

rramkumar1 replied:
In the same file as the interface definition is fine.

krzykwas replied:
Done.

rramkumar1 commented on pkg/loadbalancers/utils.go:
```diff
@@ -0,0 +1,36 @@
+	separator = ","
+)
+
+// splitAnnotation splits annotation by separator and trims whitespace
```
Can you put this in pkg/utils instead?

krzykwas replied:
Done.

rramkumar1 commented on pkg/loadbalancers/certificates.go:
```diff
@@ -82,46 +114,89 @@ func (l *L7) checkSSLCert() error {
+}
+
+// getManagedCertificates fetches SslCertificates specified via managed-certificates annotation.
+func (l *L7) getManagedCertificates() ([]*compute.SslCertificate, error) {
```
Is there a unit test for this?

krzykwas replied:
Last 3 test cases in loadbalancer_test.go test managed certificates, managed certificates + k8s secrets, managed certificates + pre-shared-cert.

rramkumar1 replied:
Do you mind adding a test for managed certificates + k8s secrets + pre-shared-cert?

rramkumar1 replied:
It would also be great if you could add a transition test (i.e add managed cert but then remove it & verify behavior)

krzykwas replied:
Done both.

bowei commented:
can you rebase and squash the commits, then split into:

- vendor/ only changes
- files changed for vendoring
- implementation
- unit tests

krzykwas commented:
I rebased, squashed and split the commits into vendor/+Gopkg.lock changes, implementation and unit tests. I didn't get what files changed for vendoring means.

rramkumar1 commented:
/cc @prameshj

rramkumar1 commented on pkg/loadbalancers/certificates.go:
```diff
@@ -23,23 +23,55 @@ import (
 func (l *L7) checkSSLCert() error {
-	// Handle Pre-Shared cert and early return if used
-	if used, err := l.usePreSharedCert(); used {
+	// Handle annotation managed-certificates
```
So the previous behavior was that if the user specified both pre-shared certs and secrets, we would only take the pre-shared certs and ignore the secrets. Correct me if I am wrong, but here it seems like you accept all three (managed, pre-shared and secret). 

I would prefer to stick with the existing semantics. Specifically, Managed certs should take precedence over everything else.

krzykwas replied:
I think that the current code either uses managed certificates and pre-shared-cert or managed certificates and k8s secrets. I think it does not support all 3 types at once. This is because of return at line 51.:
```
if used {
	l.sslCerts = append(managedSslCerts, preSharedSslCerts...)
	return err
}
```
Also I added a test case for that to loadbalancer_test.go (one of last two test cases).

prameshj replied:
Any reason we want to mix the modes? If we support both managed certs and preshared together, user might be confused why preshared and secrets don't work together or why all 3 are not picked up. Supporting just one mode with a priority in case more than one is specified might be clearer?

prameshj replied:
I guess one reason to combine preshared and managed certs is because neither of them is ingress created and hence do not need to be garbage-collected. I am curious to see if you had any other reasons, and also motivation for supporting managed certs and secrets .

krzykwas replied:
Only with modes combined it is possible to support no-downtime migration scenarios. GCLB has its own logic for selecting a certificate and Ingress really should just pass all the certificates down to GCLB. But pre-shared-cert and secrets are already implemented differently and I just wanted to keep this behavior unchanged.

rramkumar1 replied:
As discussed offline, I think it would make sense to keep consistency with the existing behavior. Namely, only take one form of cert. In this case, the precedence would be managed cert > pre-shared > secret.

krzykwas commented on pkg/utils/annotation.go:
```diff
@@ -0,0 +1,38 @@
+	separator = ","
+)
+
+// SplitAnnotation splits annotation by separator and trims whitespace
```
I don't have any strong opinion on this one and can inline this function if you think this would be better here. The arguments to have it separately are:
- this is a piece of logic used in two places - here it is separately tested, it would be harder to do it if it was interleaved with other code,
- to me it is easier to understand what the call sites of SplitAnnotations do if the parsing logic is extracted,
- it wasn't obvious to me what strings.Split("", ";") does. Namely, it returns []string{""} instead of []string{} or nil. In this case it doesn't really matter, because the annotation is being tested for being non-empty before SplitAnnotation is called, but if an empty annotation somehow gets through, the code handling pre-shared-cert in my current implementation would try to fetch SslCertificate called "", and that would generate an Event.

rramkumar1 replied:
Fair enough, lets leave it like this.

prameshj commented on pkg/events/events.go:
```diff
@@ -0,0 +1,32 @@
+/*
+Copyright 2015 The Kubernetes Authors.
```
nit : Change to 2018?

krzykwas replied:
ACK - please leave it unresolved and once we agree what changes are necessary I'll take a look to fix it.

prameshj commented on pkg/loadbalancers/certificates.go:
```diff
@@ -135,28 +210,32 @@ func getMapfromCertList(certs []*compute.SslCertificate) map[string]*compute.Ssl
+// getExistingSecretsSslCerts fetches SslCertificate resources created and managed by this load balancer
+// instance. These SslCertificate resources were created based on kubernetes secrets in Ingress
+// configuration.
+func (l *L7) getExistingSecretsSslCerts() ([]*compute.SslCertificate, error) {
```
Maybe call this "getIngressManagedSslCerts()" ?

krzykwas replied:
ACK, I can rename later when introducing all required changes.

prameshj commented on pkg/utils/annotation.go:
```diff
@@ -0,0 +1,38 @@
+	var result []string
+	for _, token := range strings.Split(annotation, separator) {
+		if token != "" {
+			result = append(result, strings.TrimSpace(token))
```
Is it possible for token to be " ", in which case we will still append an empty string to result since trimspace will make it "" ?
Maybe just check the result of TrimSpace for empty string?

krzykwas replied:
Good catch, thanks. I'll also fix it once we agree what changes are necessary.

prameshj commented on pkg/utils/annotation_test.go:
```diff
@@ -0,0 +1,42 @@
+		{"foo", []string{"foo"}},
+		{" foo ", []string{"foo"}},
+		{"foo,bar", []string{"foo", "bar"}},
+		{" foo , bar ", []string{"foo", "bar"}},
```
For the case pointed out in the above comment, maybe we can add a test case - "foo, , bar"

krzykwas replied:
ACK, will do.

krzykwas replied:
Done.

prameshj commented on pkg/loadbalancers/l7.go:
```diff
@@ -50,6 +54,10 @@ type L7RuntimeInfo struct {
+	// Ingress is the processed Ingress API object.
+	Ingress *extensions.Ingress
+	// ManagedCertificates is a comma-separated list of managed SSL certificates to use.
+	ManagedCertificates string
```
Most likely I am not understanding this fully.  If ManagedCertificate names are porvided as config, are we adding a client to watch for managed certs just to track deletions?

krzykwas replied:
pre-shared-cert is a collection of names of GCP SslCertificate resources to use for SSL termination.

ManagedCertificates are one level of indirection above. The annotation managed-certificates gives a list of names of ManagedCertificate CR. Each of them in its status provides a name of SslCertificate to use. Client is needed to fetch the objects and get the statuses.

rramkumar1 commented:
/lgtm
/hold

@prameshj if this looks good to you, we can merge.

prameshj commented on pkg/loadbalancers/certificates.go:
```diff
@@ -82,46 +108,89 @@ func (l *L7) checkSSLCert() error {
+	for _, mcrt := range mcrts {
+		if mcrt.Status.CertificateName != "" {
+			names = append(names, mcrt.Status.CertificateName)
+		}
```
Is it valid to have managed cert status with empty Certificate name? Should we log an error in this case?

krzykwas replied:
When a user creates a ManagedCertificate, he wouldn't usually fill in the status field, so the CertificateName field can be empty and it is not an error.

krzykwas commented on pkg/loadbalancers/certificates.go:
```diff
@@ -23,23 +23,49 @@ import (
+}
+
+// createSecretsSslCertificates creates SslCertificates based on kubernetes secrets in Ingress configuration.
+func (l *L7) createSecretsSslCertificates(existingCerts []*compute.SslCertificate) ([]*compute.SslCertificate, error) {
```
Done.

krzykwas commented on pkg/loadbalancers/certificates.go:
```diff
@@ -82,46 +108,89 @@ func (l *L7) checkSSLCert() error {
+	}
+
+	sel := labels.NewSelector()
+	sel.Add(*req)
```
I think it can't based on the implementation of apimachinery labels, and also based on the Go convention for constructors I observed: err != nil or err == nil and an object is instantiated.

rramkumar1 commented:
/lgtm
/hold cancel

fproulx-eoscanada commented:
@rramkumar1 when can we expect this to be available in prod ?
Also, how and when are those releases made available to Kubernetes Engine, do I need to wait for an update of my Master for it to be available ?

fproulx-eoscanada commented:
@krzykwas when can we expect this to be available in prod ?
Also, how and when are those releases made available to Kubernetes Engine, do I need to wait for an update of my Master for it to be available ?
